keps/sig-scheduling/5004-dra-extended-resource/README.md

```diff
@@ -0,0 +1,995 @@
+Allocatable.DynamicResources is updated based on DRA `DeviceClass` objects.
+
+The existing 'noderesources' plugin needs to be modified, such that a pod's
+extended resource request is checked against a NodeInfo's ScalarResources if the
+node uses device plugin, and checked against a NodeInfo's DynamicResources, if
+the request is for extended resources backed by DRA, then 'noderesources' plugin
+would pass and leave it to 'dynamicresource' plugin to check if it can be satisfied.
+
+The existing 'dynamicresources' plugin needs to be modified to account for the
+extended resource backed by DRA requests.
+
+#### EventsToRegister
+This registers all cluster events that might make an unschedulable pod schedulable,
+like finishing the allocation of a claim, or resource slice updates.
+
+The existing dynamicresurce plugin has registered all the events needed or
+extended reource backed by DRA, hence no change is needed.
+
+#### PreFilter
+It checks if the pod has any container requests for extended resources backed by DRA.
+If not, and no claims in the pod, then the plugin can return early, as there
+is nothing to do.
+
+If the pod still needs to be considered by the plugin, then it creates a
+special resource claim that has empty spec. The exact spec needs to be decided
+during Filter phase, as some node may have device plugin provide the capacity
+for the extended resource, some other node may have DRA provide the capacity.
+The requsts in the special resource claim need to vary for each node.
+
+#### Filter
+If a pod has extended resource backed by DRA, and the node does *not* have
+device plugin to provide the capacity for the resource, then the
+dynamicresource plugin needs to try allocate the resource by filling in the
+special claim's `Spec.Devices.Requests` field.
+
+One `request` is created per container, and per extended resource backed by DRA
+in the container. The `DeviceClass` in the request is randomly picked if there
+are multiple device classes advertising the extended resource. The `Name` of the
+request is determined by the container name and the extended resource name, in
+case the name conflicts with an existing device request, the next choice is
+deterministically picked until there is no conflict.
+
+The allocator needs to be modified to allow for the special resource claim for
+extended resource backed by DRA, which could vary by node. The `Allocate`
+method takes the claim as a parameter, in adddition to node parameter. The
+algorithm uses the passed in special claim whenever it processes the last claim
+in the claims slice, which is an instantiation of the special claim template
+created during the preFilter phase.
+
+If there is an allocation for a node, the allocation, and the claim are
+recorded in cyclestate.
+
+#### Score
+If two nodes can fit the pod, one node has installed device plugin, the other
+has installed DRA, score the DRA node higher. This is not going to be
+considered in ALPHA.
```

Comment:
I'm not sure whether this scoring is useful.

If both device plugin and DRA driver provide access to the same kind of hardware, then one method is not necessarily "better" than the other.

Or is the intent to automatically migrate workloads off device plugins? Then state that as the motivation here.

Reply:
my intention is to automatically migrate workloads off device plugins, but i agree we should start simple, and iterate later in the future. hence, removed the whole scorer section for now.